src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExportedMethodsWriter.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+using ILCompiler.DependencyAnalysisFramework;
+using Internal.TypeSystem;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    internal class ExportedMethodsWriter
+    {
+        private string _exportsFilePath;
+        private string _moduleName;
+        private TargetDetails _targetDetails;
+        private static ModuleDesc _systemModule;
+
+        private ExportedMethodsWriter(string outputFile, NodeFactory factory)
+        {
+            string directory = Path.GetDirectoryName(outputFile);
+            string filename = Path.GetFileNameWithoutExtension(outputFile);
+
+            _targetDetails = factory.Target;
+            _systemModule = factory.TypeSystemContext.SystemModule;
+            _moduleName = filename;
+            _exportsFilePath = Path.Combine(directory, filename + GetExportsFileExtenstion());
+        }
+
+        public static void EmitExportedMethods(string outputFile, NodeFactory factory)
+        {
+            // Only generate an export file if we're in native library mode
+            // marked by the presense of __managed__Startup method
+            if (!factory.NodeAliases.Values.Contains("__managed__Startup"))
+                return;
+
+            var nativeCallables = factory.NodeAliases.Where(n => n.Key is IMethodNode)
```

Comment:
We should get this list in ExportedMethodsRootProvider as it is rooting the native exports instead of trying to reverse engineer it here.

Reply:
Will an internal static field on `ExportedMethodsRootProvider ` suffice?

Reply:
It should not be static field. It should be instance field.

Maybe the function to produce the exports file can be moved to ExportedMethodsRootProvider  - it does not need to be in separate type.